content/local/writer.go

```diff
@@ -129,27 +123,54 @@ func (w *writer) Commit(ctx context.Context, size int64, expected digest.Digest,
 		return err
 	}
 
-	// clean up!!
-	defer os.RemoveAll(w.path)
-
 	if _, err := os.Stat(target); err == nil {
 		// collision with the target file!
+		if err := os.RemoveAll(w.path); err != nil {
+			log.G(ctx).WithField("ref", w.ref).WithField("path", w.path).Errorf("failed to remove ingest directory")
+		}
 		return errors.Wrapf(errdefs.ErrAlreadyExists, "content %v", dgst)
 	}
+
+	if w.s.ls != nil && base.Labels != nil {
+		if err := w.s.ls.Set(dgst, base.Labels); err != nil {
+			return err
+		}
+	}
+
 	if err := os.Rename(ingest, target); err != nil {
+		if w.s.ls != nil && base.Labels != nil {
+			if err := w.s.ls.Set(dgst, map[string]string{}); err != nil {
```

Comment:
Could this clear existing labels?

Reply:
Updated, this is a weird case because there is no real way to make them atomic. This label store is only used for testing and standalone right now though. I moved it back down to the end as part of setting the metadata. I think the original set wouldn't be valid if the content already existed since set completely sets the whole label set.